File: benchmarks/gemm/naive.jl
include("../../timing/dphpc_timing.jl")

const alpha = 1.5
const beta = 1.2

function init_matrices(N, M, K)

    A = zeros(Float64, N, K)
    B = zeros(Float64, K, M)
    C = zeros(Float64, N, M)

    A = [(i*j+1) % K / K for i in 1:N, j in 1:K]
    B = [(i*j+1) % M / M for i in 1:K, j in 1:M]
    C = [(i*j+1) % M / M for i in 1:N, j in 1:M]

    return A, B, C

end


function gemm(N, M, K, A, B, C)
    for i in 1:N
        for j in 1:M
            C[i, j] *= beta
            for k in 1:K
                C[i, j] += alpha * A[i, k] * B[k, j]
            end
        end
    end
end


function main()

    benchmarks = NPBenchManager.get_parameters("gemm")
    for (preset, sizes) in benchmarks
        N, M, K = collect(values(sizes))
        @dphpc_time(
            (A, B, C) = init_matrices(N,M,K), 
            gemm(N, M, K, A, B, C), 
            preset
        )
    end

end

main()

File: benchmarks/gemm/naive.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "../../timing/dphpc_timing.h"

// similar to polybench implemenation 
// Note: only works with small size arrays 


// gemm: https://docs.nvidia.com/deeplearning/performance/dl-performance-matrix-multiplication/  


#define alpha 1.5 
#define beta 1.2


// initialize matrices A, B, and C with random values
void init_matrices(int N, int M, int K, double A[N][K], double B[K][M], double C[N][M]){
    for (int i = 0; i < N; i++)
        for (int j = 0; j < K; j++)
            A[i][j] = (double) ((i*j+1) % K) / K;
    
    for (int i = 0; i < K; i++)
        for (int j = 0; j < M; j++)
            B[i][j] = (double) ((i*j+1) % M) / M;

    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++)
            C[i][j] = (double) ((i*j+1) % M) / M;

}

void gemm(int N, int M, int K, double A[N][K], double B[K][M], double C[N][M]){

    int i, j, k;
    for (i = 0; i < N; i++) {
        for (j = 0; j < M; j++) {
            C[i][j] *= beta;
            for (k = 0; k < K; k++) {
                C[i][j] += alpha * A[i][k] * B[k][j];
            }
        }
    }

}


void run_bm(int N, int M, int K, const char* preset) {

    
    double (*A)[N][K];
    double (*B)[K][M];
    double (*C)[N][M]; 

    A = (double(*)[N][K]) malloc(N*K*sizeof(double));
    B = (double(*)[K][M]) malloc(K*M*sizeof(double));
    C = (double(*)[N][M]) malloc(N*M*sizeof(double));


    dphpc_time3(
       init_matrices(N, M, K, *A, *B, *C),
       gemm(N, M, K, *A, *B, *C),
        preset
    );

    free(A);
    free(B);
    free(C); 
}

#include "_parameters.h"
int main(int argc, char** argv){

    const char *presets[] = {"S", "M", "L", "paper"};

    for (int i = 0; i < 4; i++) {
            const char* preset = presets[i];
            int n = get_params(preset)[0];
            int m = get_params(preset)[1];
            int k = get_params(preset)[2];
            run_bm(n, m, k, preset);
        }
    
    return 0;


}